### PR TITLE
Minimize the semantics of urgency

### PR DESCRIPTION
**Pull request by kazuho**
Based on our side meeting during QUIC interim, this PR makes following changes:
* Reduce semantics associated to urgency levels. Clients can now use the 8 levels in the way they like, with the only restrictions being: default=3, background=7.
* Stop using upper-case keywords for how the responses should be prioritized, or how server-provided parameters should be incorporated.

### CODE REVIEW

**Comment by kazuho** [comment]
@ianswett @afrind @LPardue @martinthomson I'd appreciate it if you could check that the changes  reflect what we discussed.

File: draft-ietf-httpbis-priority.md [hunk 1]
```diff
@@ -185,136 +185,71 @@ In both cases, the set of priority parameters is encoded as a Structured Headers
 Dictionary ({{!STRUCTURED-HEADERS}}).
 
 This document defines the urgency(`u`) and incremental(`i`) parameters. When
-used, these parameters MUST be accompanied by values. When any of the defined
-parameters are omitted, or if the Priority header field is not used, their
-default values SHOULD be applied.
+receiving an HTTP request that does not carry these priority parameters, a
+server SHOULD act as if the default values were specified.
```
Comment (LPardue) [inline draft-ietf-httpbis-priority.md]: I see that the HTTP response case is mentioned later but I think it might help to explicitly mention that it here e.g. a response with the parameters SHOULD NOT act be interpreted as if default were specified.

File: draft-ietf-httpbis-priority.md [hunk 2]
```diff
@@ -185,136 +185,71 @@ In both cases, the set of priority parameters is encoded as a Structured Headers
...
-to be rendered by that font.
-
-### supplementary
-
-The supplementary urgencies (values 2 to 6) indicate a response that is helpful
-to the client using a composition of responses, even though the response itself
-is not mandatory for using those responses.
+A client that fetches a document that likely consists of multiple HTTP resources
+(e.g., HTML) SHOULD assign the default urgency level to the main resource.  This
+convention provides the servers the capability of refining the urgency using
```
Comment (LPardue) [inline draft-ietf-httpbis-priority.md]:
```suggestion
convention allows servers to refine the urgency using
```
suggestion to make the text more assertive

File: draft-ietf-httpbis-priority.md [hunk 3]
```diff
@@ -473,9 +408,15 @@ HTTP responses deserve to be prioritized. For example, use of an HTML document
 might depend heavily on one of the inline images. Existence of such
 dependencies is typically best known to the server.
 
-By using the "Priority" response header, a server can override the
-prioritization hints provided by the client. When used, the parameters found
-in the response header field overrides those specified by the client.
+An origin can use the Priority response header field to indicate its view on how
+an HTTP response should be proioritized. When forwarding an HTTP response with
```
Comment (LPardue) [inline draft-ietf-httpbis-priority.md]:
```suggestion
an HTTP response should be prioritized. When forwarding an HTTP response with
```

File: draft-ietf-httpbis-priority.md [hunk 4]
```diff
@@ -495,7 +436,7 @@ content-type = image/png
 priority = u=2
 ~~~
 
-the intermediary's understanding of the urgency is promoted from `4` to `2`,
+the intermediary might alter its understanding of the urgency from `4` to `2`,
```
Comment (LPardue) [inline draft-ietf-httpbis-priority.md]:
```suggestion
the intermediary might alter its understanding of the urgency from `5` to `2`,
```

File: draft-ietf-httpbis-priority.md [hunk 5]
```diff
@@ -185,136 +185,72 @@ In both cases, the set of priority parameters is encoded as a Structured Headers
...
-a precondition for the user to use that portion of the HTML document, which is
-to be rendered by that font.
-
-### supplementary
-
-The supplementary urgencies (values 2 to 6) indicate a response that is helpful
-to the client using a composition of responses, even though the response itself
-is not mandatory for using those responses.
+A client that fetches a document that likely consists of multiple HTTP resources
+(e.g., HTML) SHOULD assign the default urgency level to the main resource.  This
```
Comment (tombergan) [inline draft-ietf-httpbis-priority.md]: Sorry I missed the meeting. Can you explain the rationale for this? Since there is no resource more urgent than the main resource, this effectively reserves priorities 0-2 for the server.

Comment (kazuho) [reply draft-ietf-httpbis-priority.md]: Sorry if you felt left alone. I have written down my understanding of the outcome of the meeting at https://github.com/httpwg/http-extensions/issues/1026#issuecomment-582776695, I think we might want to consider discussing there.

Comment (tombergan) [reply draft-ietf-httpbis-priority.md]: No worries and thanks for the context. The examples make sense. I think it's fine to say that 7 should be reserved for background downloads and it's fine to say that the main HTML should use level X, for some X, but I'm not sure why X=3 was chosen. It seems like this could work just as well with X=0 or X=1.

Note that Chrome (and WebKit/Safari?) use 6 priority levels currently. With X = 3, there are 5 levels available which means that Chrome (and WebKit/Safari?) need to fold two levels into one. This is not a dealbreaker but IMO there should at least be a good reason why levels 0-2 need to be reserved for the server when clients have a demonstrated use for 6 of 8 levels. Ideally someone from Firefox (and Safari) should chime in as well.

It's also worth mentioning that this pull request doesn't have enough information to do everything that [this comment](https://github.com/httpwg/http-extensions/issues/1026#issuecomment-582776695) wants to do. For example: "a server might demote them to the same level as asynchronously fetched JavaScript files, or to somewhere even below, but above background downloads." It's not possible to demote to the same level as async JS, because the level of async JS is not specified. Also, it's not trivial to set a priority "above background downloads" because the pull request merely states that level 7 is reserved for background downloads, not anything about level 6 (perhaps some client has many background downloads and wants to partition those into two levels). I wouldn't change this pull request though -- this pull request seems like the right level of detail.

To summarize, this looks good but I would suggest changing the level of mainframe requests to 0 or 1.

File: draft-ietf-httpbis-priority.md [hunk 6]
```diff
@@ -185,136 +185,72 @@ In both cases, the set of priority parameters is encoded as a Structured Headers
...
-
-### background
-
-The background urgency (value 7) is used for responses of which the delivery can
-be postponed without having an impact on using other responses.
-
-As an example, the download of a large file in a web browser would be assigned
-the background urgency so it would not impact further page loads on the same
-connection.
+The lowest urgency level (7) is reserved for background tasks such as delivery
```
Comment (tombergan) [inline draft-ietf-httpbis-priority.md]: It might be worth specifying which urgency level should be used for server push streams. I think the "optimal" solution is to default to 7, then rely on the client to reprioritize the pushed stream when it matches an actual requests. Assuming no bufferbloat, this ensures that the server does not waste any high-priority bandwidth on pushes that won't be used.

Comment (afrind) [reply draft-ietf-httpbis-priority.md]: There are non-browser use-cases of push where background might not be the best level for pushed resources, but as long as we are not too prescriptive about it I don't object to saying that background is recommended.

Comment (LPardue) [reply draft-ietf-httpbis-priority.md]: I think @tombergen's suggestion is a sensible one. We can support @afrind case by applying the server signal to the push; absence of server signal means default (in this case background), presence of signal prioritizes at a different level and, if passed back to the client, is an explicit signal of the push priority. If this matches the priority the client would have reprioritized to (because it is more important) you save the client from sending a pointless reprioritization.

File: draft-ietf-httpbis-priority.md (hunk 5 above)
Comment (kazuho) [reply draft-ietf-httpbis-priority.md]: Thank you for sharing your thoughts.

Regarding background, you are right! I hadn't thought about that, but the new design does not forbid a client from using level(s) above 7 for background tasks too. That said, I agree with you that we do not need to be more strict.

Regarding the default level, based on your comments, I am going to change it to 1. I think it would be a good idea to provide one level above default, considering the fact that Firefox prioritizes JavaScript and CSS files above HTML. But I do not think we need more than one.

Then, if people prefer having symmetry (i.e. having same number of urgency levels above and below the default), we can consider increasing the number of levels to 10 (or maybe 12) so that there would be sufficient number of levels below the default.

File: draft-ietf-httpbis-priority.md (hunk 6 above)
Comment (kazuho) [reply draft-ietf-httpbis-priority.md]: IMO, the default should (can) be 7 only when all the following conditions are met:
* 1. There is no bufferbloat.
* 2. If the specification retains the definition of a reprioritization frame.
* 3. If the browser correctly uses that.

For point 1, while there would likely be no bufferbloat in HTTP/3, there will be in HTTP/2. For point 2, the discussion is ongoing in #1021, and I am not sure if we can be certain that browsers would reprioritize pushed assets even in case of it being specified.

Considering these aspects, I tend to think that the correct thing for a server to do by default would be prioritize the pushed response right after the response that initiated the push. In terms of the specification, we can either state to use the same urgency level with i=false, or the next urgency level.

Such a design would give us what we have now in HTTP/2.

In case of HTTP/3, there would be negligible negative impact, assuming that there would be no bufferbloat. This is because URLs of pushed responses and the HTML would reach clients before the pushed resources do. Clients would be possible to cancel / reprioritize the pushed resources when or before when it discovers and issues requests for the non-pushed resources that it needs.

Comment (tombergan) [reply draft-ietf-httpbis-priority.md]: I think it's fine to punt this question to #1021. You are right that this becomes problematic without reprioritization. If repriotization goes away, perhaps we'll need to say something like "servers should not push resources unless they are confident they can assign a suitable priority for that resource". 

> I tend to think that the correct thing for a server to do by default would be prioritize the pushed response right after the response that initiated the push

We have observed that H2 push can be a serious footgun. It's easy to hurt performance by over-pushing, such as by pushing low priority resources or cached resources. A level 7 default reduces the potential impact of this footgun. As @LPardue said, this is just a default; the server can always do something different when it believes it knows better.

> Clients would be possible to cancel the pushed resources when or before when it discovers

It's difficult for the client to know whether or not they should cancel a pushed resource. Maybe they will need the resource soon, they just don't know yet.

> I am not sure if we can be certain that browsers would reprioritize pushed assets even in case of it being specified.

FWIW, Chrome already does this for H2. Not sure about other browsers.

Comment (kazuho) [reply draft-ietf-httpbis-priority.md]: Created #1056 so that we do not miss it.

File: draft-ietf-httpbis-priority.md (hunk 5 above)
Comment (kazuho) [reply draft-ietf-httpbis-priority.md]: Done in bc04fca.

File: draft-ietf-httpbis-priority.md (hunk 1 above)
Comment (kazuho) [reply draft-ietf-httpbis-priority.md]: Done in 146fd30. PTAL.

**Comment by kazuho** [comment]
@LPardue I think this PR is becoming ready. Would you mind taking a look (and merge it if looks okay to you)?
